pkg/controller/storageprofile-controller.go

```diff
@@ -0,0 +1,200 @@
+	log := r.log.WithValues("StorageProfile", req.NamespacedName)
+	log.Info("reconciling StorageProfile")
+
+	storageClass := &storagev1.StorageClass{}
+	if err := r.uncachedClient.Get(context.TODO(), req.NamespacedName, storageClass); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return r.reconcileStorageProfile(storageClass)
+}
+
+func (r *StorageProfileReconciler) reconcileStorageProfile(sc *storagev1.StorageClass) (reconcile.Result, error) {
+	log := r.log.WithValues("StorageProfile", sc.Name)
+
+	storageProfile, err := r.createStorageProfile(sc)
+	if err != nil {
+		log.Error(err, "Unable to create StorageProfile")
+		return reconcile.Result{}, err
+	}
+	currentStorageProfile := storageProfile.DeepCopyObject()
+
+	storageProfile.Status.StorageClass = &sc.Name
+	storageProfile.Status.Provisioner = &sc.Provisioner
+	storageProfile.Status.ClaimPropertySets = storageProfile.Spec.ClaimPropertySets
+
+	var claimPropertySet *cdiv1.ClaimPropertySet
+	// right now the CDI supports only a single property set
+	if len(storageProfile.Status.ClaimPropertySets) > 0 {
+		claimPropertySet = &storageProfile.Status.ClaimPropertySets[0]
+	} else {
+		claimPropertySet = &cdiv1.ClaimPropertySet{}
+	}
+
+	r.reconcileAccessModes(sc, claimPropertySet)
+	r.reconcileVolumeMode(sc, claimPropertySet)
+
+	if !isClaimPropertySetEmpty(claimPropertySet) {
+		storageProfile.Status.ClaimPropertySets = []cdiv1.ClaimPropertySet{*claimPropertySet}
+	}
+
+	if !reflect.DeepEqual(currentStorageProfile, storageProfile) {
+		// Updates have happened, update StorageProfile.
+		log.Info("Updating StorageProfile", "StorageProfile.Name", storageProfile.Name, "storageProfile", storageProfile)
+		if err := r.client.Update(context.TODO(), storageProfile); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func isClaimPropertySetEmpty(set *cdiv1.ClaimPropertySet) bool {
+	return set == nil ||
+		(len(set.AccessModes) == 0 && set.VolumeMode == nil)
+}
+
+func (r *StorageProfileReconciler) reconcileVolumeMode(sc *storagev1.StorageClass, claimPropertySet *cdiv1.ClaimPropertySet) {
+	if claimPropertySet.VolumeMode == nil {
+		capabilities, found := storagecapabilities.Get(sc)
+		if found {
+			claimPropertySet.VolumeMode = &capabilities.VolumeMode
+		}
+	}
+}
+
+func (r *StorageProfileReconciler) reconcileAccessModes(sc *storagev1.StorageClass, claimPropertySet *cdiv1.ClaimPropertySet) {
+	// reconcile accessModes
+	if len(claimPropertySet.AccessModes) == 0 {
+		capabilities, found := storagecapabilities.Get(sc)
+		if found {
+			claimPropertySet.AccessModes = []v1.PersistentVolumeAccessMode{capabilities.AccessMode}
+		}
+	}
+}
+
+func (r *StorageProfileReconciler) createStorageProfile(sc *storagev1.StorageClass) (*cdiv1.StorageProfile, error) {
+	storageProfile := &cdiv1.StorageProfile{}
+	if err := r.uncachedClient.Get(context.TODO(), types.NamespacedName{Name: sc.Name}, storageProfile); err != nil {
```

Comment:
why uncached, this is cluster scoped, right?

Reply:
the same as in previous comment